common/containers/Tabs/GenerateWallet/components/Mnemonic/Mnemonic.tsx

```diff
@@ -0,0 +1,150 @@
+    const { words, confirmValues, isConfirming, isConfirmed } = this.state;
+    let content;
+
+    if (isConfirmed) {
+      content = <FinalSteps walletType={WalletType.Mnemonic} />;
+    } else {
+      const canContinue = this.checkCanContinue();
+      const firstHalf: WordTuple[] = [];
+      const lastHalf: WordTuple[] = [];
+      words.forEach((word, index) => {
+        if (index < words.length / 2) {
+          firstHalf.push({ word, index });
+        } else {
+          lastHalf.push({ word, index });
+        }
+      });
+
+      content = (
+        <div>
+          <h1 className="GenerateMnemonic-title">Generate a {translate('x_Mnemonic')}</h1>
+
+          <p className="GenerateMnemonic-help">
+            {isConfirming
+              ? `
+              Re-enter your phrase to confirm you copied it correctly. If you
+              forgot one of your words, just click the button beside the input
+              to reveal it.
+            `
+              : `
+              Write these words down. Do not copy them to your clipboard, or save
+              them anywhere online.
+            `}
+          </p>
+
+          <div className="GenerateMnemonic-words">
+            {[firstHalf, lastHalf].map((ws, i) => (
+              <div key={i} className="GenerateMnemonic-words-column">
+                {ws.map(word => (
+                  <Word
+                    key={`${word.word}${word.index}`}
+                    index={word.index}
+                    word={word.word}
+                    value={confirmValues[word.index] || ''}
+                    isReadOnly={!isConfirming}
+                    onChange={this.handleConfirmChange}
+                  />
+                ))}
+              </div>
+            ))}
+          </div>
+
+          <div className="GenerateMnemonic-buttons">
+            {!isConfirming && (
+              <button
+                className="GenerateMnemonic-buttons-btn btn btn-default"
+                onClick={this.regenerateWordArray}
+              >
+                <i className="fa fa-refresh" /> Regenerate Phrase
+              </button>
+            )}
+            <button
+              className="GenerateMnemonic-buttons-btn btn btn-primary"
+              disabled={!canContinue}
+              onClick={this.goToNextStep}
+            >
+              Confirm Phrase
+            </button>
+          </div>
+
+          <button className="GenerateMnemonic-skip" onClick={this.skip} />
+        </div>
+      );
+    }
+
+    return <div className="GenerateMnemonic Tab-content-pane">{content}</div>;
+  }
+
+  private regenerateWordArray = () => {
+    this.setState({ words: generateMnemonic().split(' ') });
+  };
+
+  private handleConfirmChange = (index: number, value: string) => {
+    console.log(index, value);
+    const confirmValues = [...this.state.confirmValues];
+    confirmValues[index] = value;
+    console.log(confirmValues);
+    this.setState({ confirmValues });
```

Comment:
This could all be done with a setState cb that'll give you the previous state. Should be more clean.